### PR TITLE
feat: use repo default branch for template resolution

### PR DESCRIPTION
**Pull request by wass3r**
closes https://github.com/go-vela/community/issues/134

See the issue for more details. The proposed changes should result in Vela pulling from the default branch that Vela is aware of from when the Repo was added or hook for the repo was created.

### CODE REVIEW

File: registry/github/parse_test.go [hunk 1]
```diff
@@ -36,7 +36,9 @@
 	}
 
 	path := "github.example.com/github/octocat/template.yml"
-	got, err := c.Parse(path)
+	defaultBranch := ""
+
+	got, err := c.Parse(path, defaultBranch)
```
[the remark this answers is not in the document]
Comment (wass3r) [reply registry/github/parse_test.go:41]: i think in a separate PR we should exclude `dupl` from test files.

File: registry/github/parse_test.go [hunk 2]
```diff
@@ -46,7 +48,39 @@
...
+	c, err := New(s.URL, "")
+	if err != nil {
+		t.Errorf("Creating client returned err: %v", err)
+	}
+
+	path := "github.example.com/github/octocat/template.yml"
+	defaultBranch := "master"
+
+	got, err := c.Parse(path, defaultBranch)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (wass3r) [reply registry/github/parse_test.go:75]: i think in a separate PR we should exclude `dupl` from test files.

File: compiler/native/expand.go [hunk 3]
```diff
@@ -55,7 +55,7 @@ func (c *client) ExpandSteps(s yaml.StepSlice, tmpls map[string]*yaml.Template)
 		}
 
 		// parse source from template
-		src, err := c.Github.Parse(tmpl.Source)
+		src, err := c.Github.Parse(tmpl.Source, c.repo.GetBranch())
```
Comment (wass3r) [inline compiler/native/expand.go:58]: we're grabbing default branch from the repo in the current context and passing it along to `Parse()`

File: registry/github/parse.go [hunk 4]
```diff
@@ -33,8 +36,13 @@ func (c *client) Parse(path string) (*registry.Source, error) {
...
 
-	// set the default reference to master
-	ref := "master"
+	// use the fallback default reference
+	ref := defaultRef
+
+	// override with supplied defaultBranch
+	if len(defaultBranch) > 0 {
+		ref = defaultBranch
+	}
```
Comment (wass3r) [inline registry/github/parse.go:45]: here's the `meat` of the change. the fallback branch is `main` but should normally use the default branch associated with the repo in the context.

File: registry/registry.go [hunk 5]
```diff
@@ -10,8 +10,9 @@ import "github.com/go-vela/types/library"
 // with the different supported template registries.
 type Service interface {
 	// Parse defines a function that creates the
-	// registry source object from a template path.
-	Parse(string) (*Source, error)
+	// registry source object from a template path
+	// and default branch.
+	Parse(string, string) (*Source, error)
```
Comment (wass3r) [inline registry/registry.go:15]: updating the interface to account for the new parameter

File: registry/github/parse.go [hunk 6]
```diff
@@ -12,8 +12,11 @@ import (
 	"github.com/goware/urlx"
 )
 
-// Parse creates the registry source object from a template path.
-func (c *client) Parse(path string) (*registry.Source, error) {
+const defaultRef = "main"
```
Comment (jbrockopp) [inline registry/github/parse.go:15]: Are we sure we want to make `main` the fall back branch?

I understand why the change is being made from `master` to `main` 👍 

But my concern is for all the repos that are using templates, but are still using `master` as their default branch.

What happens in that scenario?

Comment (wass3r) [reply registry/github/parse.go:15]: I think it's fine. Default branch passed in from the context should always be either `master` (current default if nothing was passed in at activation) or something the user defined (if the repo was activated via CLI or API; UI doesn't allow customization there). From what I can tell, we currently do not allow the user to activate a repo with an empty default branch, so the likelyhood of this falling back to `main` are basically nil, as far as I can tell. But it's possible I miss(ed|-understood) something.

If somehow it got this state, the template would fail to be pulled just as if someone's default is not `master` and they were trying to reference a template without `@<branch>`. I'm fine with that, personally.

However, I think it might be good to modify the docs to encourage the use of `@<branch>` identifier when referencing templates as a general good practice to be explicit.

As an aside, I do have another change in the hopper for the API to deal with https://github.com/go-vela/server/blob/master/api/repo.go#L113 and https://github.com/go-vela/server/blob/master/api/repo.go#L183

Comment (jbrockopp) [reply registry/github/parse.go:15]: Sure 👍 

I'm still a little uncomfortable with the change (setting the fallback default to `main`) because of the potential impact of breaking pipelines.

But perhaps I'm being overly cautious and this is a "rip the bandaid" scenario?

At any rate, if others don't feel strongly then I don't want to be the stick in the mud.

**Review by jbrockopp — APPROVED** [review]
LGTM

File: registry/github/parse.go (hunk 6 above)
Comment (wass3r) [reply registry/github/parse.go:15]: Thanks for raising the concern. I do think it's a "rip the bandaid" scenario, but I also think the risk is very low. I can be convinced otherwise if there's something I didn't think of or code I missed somewhere. Regardless, thanks for checking in out.

**Review by davidvader — APPROVED** [review]
LGTM, nice change
